### PR TITLE
SLVS-2054 Implement “startProgress” and “reportProgress”

### PR DESCRIPTION
**Pull request by gabriela-trutan-sonarsource**
[SLVS-2054](https://sonarsource.atlassian.net/browse/SLVS-2054)

[SLVS-2054]: https://sonarsource.atlassian.net/browse/SLVS-2054?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/SLCore/Protocol/EitherJsonConverter.cs [hunk 1]
```diff
@@ -110,6 +112,18 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                 }
             }
 
+            if (!jsonProperties.Any())
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Protocol/EitherJsonConverter.cs:121]: it would be easier to read if conditions were inverted

Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Protocol/EitherJsonConverter.cs:121]: This block can also be put before the foreach

File: src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs [hunk 2]
```diff
@@ -44,34 +44,18 @@ public void SerializeObject_SerializesEitherAsSingleObject()
...
         JsonConvert.SerializeObject(left).Should().BeEquivalentTo("""{"Property":{"Left":"lll"}}""");
         JsonConvert.SerializeObject(right).Should().BeEquivalentTo("""{"Property":{"Right":10}}""");
     }
 
-    [TestMethod]
-    public void DeserializeObject_PrimitiveNotAnObject_Throws()
-    {
-        var str = """
-                  {
-                    "Property" : "ThisIsExpectedToBeAnObjectButItIsAString"
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs]: Why was this removed? It's still true and it still passes

File: src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs [hunk 3]
```diff
@@ -201,4 +248,34 @@ public class RightOption
...
     }
+
+    public class ObjectWithRightOptionNoProperties
+    {
+        [JsonConverter(typeof(EitherJsonConverter<LeftOption, RightOption>))]
+        public Either<LeftOption, RightOption> Property { get; set; }
+
+        public class LeftOption
+        {
+            public string Left { get; set; }
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs]: Use LeftProperty instead of Left (same for Right) as it is clearer that it refers to a property of the Left type not the Left type itself. Update SimpleObject class as well

File: src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs [hunk 4]
```diff
@@ -126,13 +109,78 @@ public void DeserializeObject_UnresolvableTypes_Throws()
                 "Types SonarLint.VisualStudio.SLCore.UnitTests.Protocol.EitherJsonConverterTests+ConflictingObject+ConflictingLeft and SonarLint.VisualStudio.SLCore.UnitTests.Protocol.EitherJsonConverterTests+ConflictingObject+ConflictingRight have equivalent sets of properties and fields");
     }
 
+    [TestMethod]
+    public void DeserializeObject_ObjectWithRightOptionNoProperties_Right_ChoosesCorrectSide()
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs:163]: need a test where both left and right are empty objects

Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs:163]: and another one where they're not empty, but the values for them are omitted in json (for example, if NullValueHandling is set to Ignore)

File: src/SLCore.Listeners/Implementation/ProgressListener.cs [hunk 5]
```diff
@@ -19,32 +19,37 @@
...
-        public Task ReportProgressAsync(object parameters)
+        /// <inheritdoc />
+        public void ReportProgress(ReportProgressParams parameters)
         {
-            return Task.CompletedTask;
+            if (parameters.notification.Left is { } updateNotification)
+            {
+                slCoreStatusNotifier.Notify(FormatNotification(updateNotification.message), false);
+            }
+            else if (parameters.notification.Right != null)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.Listeners/Implementation/ProgressListener.cs:47]: Rider complains that this code is unreachable because Left is not nullable. We need to adapt the EIther class like this

```
public TLeft? Left { get; private init; }
public TRight? Right { get; private init; }
```

File: src/SLCore.Listeners/Implementation/ProgressListener.cs [hunk 6]
```diff
@@ -19,32 +19,37 @@
...
-        /// <summary>
-        /// Stub method for compability with SLCore. We do not support progress
-        /// </summary>
-        /// <param name="parameters">Parameter's here for compability we discard it</param>
-        public Task ReportProgressAsync(object parameters)
+        /// <inheritdoc />
+        public void ReportProgress(ReportProgressParams parameters)
         {
-            return Task.CompletedTask;
+            if (parameters.notification.Left is { } updateNotification)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.Listeners/Implementation/ProgressListener.cs]: Personal preference: `parameters.notification is {Left: {} updateNotification}`

File: src/SLCore.Listeners/Implementation/ProgressListener.cs [hunk 7]
```diff
@@ -19,32 +19,37 @@
...
 using SonarLint.VisualStudio.SLCore.Core;
 using SonarLint.VisualStudio.SLCore.Listener.Progress;
 
 namespace SonarLint.VisualStudio.SLCore.Listeners.Implementation
 {
     [Export(typeof(ISLCoreListener))]
     [PartCreationPolicy(CreationPolicy.Shared)]
-    public class ProgressListener : IProgressListener
+    [method: ImportingConstructor]
+    public class ProgressListener(IStatusBarNotifier slCoreStatusNotifier) : IProgressListener
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.Listeners/Implementation/ProgressListener.cs]: why is it called `slCoreStatusNotifier`?

File: src/SLCore.Listeners/Implementation/ProgressListener.cs [hunk 8]
```diff
@@ -19,32 +19,37 @@
...
+            {
+                slCoreStatusNotifier.Notify(FormatNotification(updateNotification.message), false);
+            }
+            else if (parameters.notification.Right != null)
+            {
+                slCoreStatusNotifier.Notify(string.Empty, false);
+            }
         }
+
+        private string FormatNotification(string message) => $"{SLCoreStrings.LongProductName}: {message}";
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.Listeners/Implementation/ProgressListener.cs]: Does the `LongProductName` fit well into the task bar?

File: src/SLCore.Listeners/Implementation/ProgressListener.cs [hunk 9]
```diff
@@ -19,32 +19,37 @@
...
-        /// </summary>
-        /// <param name="parameters">Parameter's here for compability we discard it</param>
-        public Task ReportProgressAsync(object parameters)
+        /// <inheritdoc />
+        public void ReportProgress(ReportProgressParams parameters)
         {
-            return Task.CompletedTask;
+            if (parameters.notification.Left is { } updateNotification)
+            {
+                slCoreStatusNotifier.Notify(FormatNotification(updateNotification.message), false);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.Listeners/Implementation/ProgressListener.cs]: Have you looked into this? https://sonarsource.atlassian.net/browse/SLVS-2054?focusedCommentId=800661 Seems like a better alternative, becuase we may have multiple progresses from slcore and they might overwrite one another

File: src/SLCore/Protocol/EitherJsonConverter.cs (hunk 1 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Protocol/EitherJsonConverter.cs:121]: I chose to put it after the foreach, because this is a special, edge case. It's the first DTO encountered that reached this case. So for performance reasons, after the foreach is better

File: src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs (hunk 2 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.UnitTests/Protocol/EitherJsonConverterTests.cs]: It was not intentional. :O

File: src/SLCore.Listeners/Implementation/ProgressListener.cs (hunk 7 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.Listeners/Implementation/ProgressListener.cs]: Because of a previous local implementation

File: src/SLCore.Listeners/Implementation/ProgressListener.cs (hunk 8 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.Listeners/Implementation/ProgressListener.cs]: It does. The notifications received from SlCore are quite short

Comment (gabriela-trutan-sonarsource) [reply src/SLCore.Listeners/Implementation/ProgressListener.cs]: It's also similar to the analysis [notifications](https://github.com/SonarSource/sonarlint-visualstudio/blob/eb9a58a375d3ddc2595768ad8c98dc8f20351464/src/Integration.Vsix/Analysis/AnalysisStrings.resx#L151)

File: src/SLCore.Listeners/Implementation/ProgressListener.cs (hunk 9 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.Listeners/Implementation/ProgressListener.cs]: I think it's a nice feature, but unfortunately it does not apply here. It shows a status while you run a task on the background, so it can be cancelled etc. 
In our case, we just receive the notification from SlCore. Of course we can create a hack to run a task and wait until we receive the notification that the task with the same id ended, but I don't think it's the intended use for this case. Also what happens if, for whatever reason, SlCore never notifies us about a task being ended? I find it too risky

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]
